### DISCUSSION TITLE
Support for anonymous identity and trait evaluations

### DISCUSSION THREAD

**Discussion by augustuswm** [Flagsmith/flagsmith#77] [Q&A]
I was looking at trying to integrate Flagsmith as the backend flag service for some of our internal projects. There were a couple things that our current tooling can do that I couldn't find something analogous for.

1. Often times with flag evaluations we want to be able to pass an identity identifier and user traits that we do not want to persist to a database. It might be a case where it is specific trait data that we do not want stored, or a cases where we have a lot of anonymous users and we do not want all of them to be persisted.
2. Getting evaluations for a specific subset of flags. Often times we only need to have a subset of the entire flag list for a product evaluated. It looks like for an identity we have to select / and return all of the flags for a environment / identity currently.

I have only been looking at the API side of things, and tried to mock up a proof of concept to see if I could get that functionality added (I'm not a Python or Django dev): https://github.com/augustuswm/flagsmith-api/commit/3ae32acda6c291c04bcb7e7c51108b41f2e165c7

Ignoring the hacked together implementation, is the concept of ad-hoc non-persisted identities / traits / evaluations something that could work within the scope of Flagsmith? or by trying to integrate this functionality am I likely to cause larger issues (data / stability)?

Thanks for any advice, and appreciate you making this open source. It looks like a great platform.

- Augustus

**Comment by dabeeeenster** [comment]
Hi Augustus,

Thanks so much for this code and design. This is an area we are working on at the moment. 100% agree it makes sense to have an 'anonymous' flag that you can send when you identify someone in Flagsmith. 

We are currently trying to work out the best design for an `merge` style function similar to the way mixpanel handles this problem (https://help.mixpanel.com/hc/en-us/articles/360039133851#enable-id-merge). Do you have any thoughts on this? It's important that when bucketing identities for segments with the % split operator, the user is bucketed consistently both before and after they authenticate (more info on how the bucketing algorithm works is here https://docs.flagsmith.com/staged-feature-rollouts/#how-does-it-work:

> Every Identity/Segment combination is combined and then hashed, and a floating point value between 0.0 and 1.0 is generated from this hash. This value is then evaluated against the "% Split" rule.

We'll take a look at the code and reply back here early next week. 

Thanks again!
Ben

**Comment by augustuswm** [comment]
What mix panel is doing is pretty interesting, and I'll need to read in to it more. I am presuming it is some kind of aliasing of multiple ids up to a main primary id. How you determine what to serve when evaluating a flag for two aliased ids that each result in different values is difficult, and I'm not sure what the correct answer would be there.

For our organization we have passed the identification responsibility off to the products that use the flag service. If they want an anonymous user to have the same flag evaluation results after they sign up, then we require that they provide the same identifier value (and relevant traits) as was used when the user was not signed up. Essentially we would only guarantee to other services that an evaluation result will remain the same for a given flag if the identity identifier, traits, and segments remain the same.

I definitely see the appeal though of having some of that functionality built in to the flag service it self.

**Comment by dabeeeenster** [comment]
Yes that's right, so if you sent an anonymously flagged identity ID initially, then send e merge request with the current anonymous ID and the authenticated ID, we would guarantee they would be in the same consistent segment bucket.

Do you mind sharing your product?

**Comment by augustuswm** [comment]
We are currently using https://github.com/markphelps/flipt (with slight authentication changes). Currently it is really only responsible for taking an identity id and some key value context data and returning an on / off evaluation. It only stores the flag and targeting data, and does not store user/identity data.

Our system for identifying users / assigning ids I can't share though as that is a custom internal project.

I'm particularly interested in Flagsmith because of the way it models Projects > Environments as well as being able to add overrides at a per user/identity level. We currently offload all of the user specific work to our identity system, which just results in confusion internally.

**Comment by Skwai** [comment]
Any progress on this? We tried using the anonymous ID to identify users in Flagsmith but from looking in Mixpanel the flags aren't being associated correctly. It looks like we still lack the ability to send an `anonymous_id` to Fagsmith identify?
